// @mapbox/node-pre-gyp/lib/package.js
'use strict';

module.exports = exports = _package;

exports.usage = 'Packs binary (and enclosing directory) into locally staged tarball';

const fs = require('fs');
const path = require('path');
const log = require('npmlog');
const versioning = require('./util/versioning.js');
const napi = require('./util/napi.js');
const makeDir = require('make-dir');
const tar = require('tar');

function readdirSync(dir) {
  let list = [];
  const files = fs.readdirSync(dir);

  files.forEach((file) => {
    const fullPath = path.join(dir, file);
    const stats = fs.lstatSync(fullPath);
    if (stats.isDirectory()) {
      list = list.concat(readdirSync(fullPath));
    } else {
      list.push(fullPath);
    }
  });
  return list;
}

function _package(gyp, argv, callback) {
  const package_json = gyp.package_json;
  const napi_build_version = napi.get_napi_build_version_from_command_args(argv);
  const opts = versioning.evaluate(package_json, gyp.opts, napi_build_version);
  const from = opts.module_path;
  const binary_module = path.join(from, opts.module_name + '.node');
  
  existsAsync(binary_module, (found) => {
    if (!found) {
      return callback(new Error(`Cannot package because ${binary_module} is missing. Run 'node-pre-gyp rebuild' first.`));
    }
    
    const tarball = opts.staged_tarball;
    const filter_func = function (entry) {
      const basename = path.basename(entry);
      if (basename.length && basename[0] !== '.') {
        log.info('packing', entry);
        return true;
      } else {
        log.verbose('skipping', entry);
      }
      return false;
    };

    makeDir(path.dirname(tarball)).then(() => {
      const files = readdirSync(from).map(file => path.join(path.basename(from), path.relative(from, file)));
      
      tar.create({
        portable: false,
        gzip: true,
        filter: filter_func,
        file: tarball,
        cwd: path.dirname(from)
      }, files, (err) => {
        if (err) {
          log.error('package', err.message);
        } else {
          log.info('package', `Binary staged at "${tarball}"`);
        }
        callback(err);
      });
    }).catch(callback);
  });
}
